### PR TITLE
Keep render tree and element tree in sync when re-used elements move in a MultiChildRenderObjectElement's child list

### LINKED ISSUE #48855: GlobalKeys causing unexpected rendered Widget position
Steps to Reproduce

1. Run flutter create
2. Add the three files at this gist to your project, including replacing main.dart: https://gist.github.com/emshack/0bdfee8f86377301075ee573a994d66c
3. Run the app on a target device
4. Observe three things: A tree structure on the left, "Expected order" text on the right, and an icon button at the bottom. The order of the tree structure from top to bottom should match the order in "Expected order" from left to right. (Nodes 0-8, where 0 is a group node containing children 1-6, and 6 is a group node containing children 7-8)
5. Click the icon button. This moves group node 6 to index 0 in group 0's children.
6. Observe that the tree structure now no longer matches the "Expected order" text. While group node 6 has moved to the correct position in the tree, its children have not moved with it, and child node 1 has moved to the bottom. The "Expected order" text, which is determined within the tree's build method and operates off the same list of widgets passed to the tree, reflects the expected tree structure.
7. Comment out either the if clause in line 113 or all of line 116 in tree.dart and do a Hot Restart. This either removes or refreshes the GlobalKeys passed into each row in the tree.
8. Observe the original tree structure and "Expected order" text.
9. Click the icon button.
10. Observe that the tree structure now appears as expected, and matches the "Expected order" text.

Any idea what's going on here? Am I violating some assumption about GlobalKeys in how I'm using them, or does this look like a Flutter bug?

Flutter Doctor
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel master, v1.13.10-pre.60, on Mac OS X 10.14.6 18G2022, locale en-US)
[!] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    ! Some Android licenses not accepted.  To resolve this, run: flutter doctor --android-licenses
[✓] Xcode - develop for iOS and macOS (Xcode 11.1)
[!] Android Studio
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
[✓] Android Studio (version 3.1)
[✓] IntelliJ IDEA Community Edition (version 2019.1.2)
[✓] VS Code (version 1.41.1)
[✓] Connected device (2 available)

### PR DESCRIPTION
**Pull request by goderbauer**
## Description

Prior to this change the Element child list of an MultiChildRenderObjectElement and the RenderObject child list of its associated RenderObject could get out of sync when Elements move around in the child list. See https://github.com/flutter/flutter/issues/48855#issuecomment-592261646 for a detailed description of the problem.

In a nutshell, we used to only change the slot of a child when its proceeding sibling changed. Updating a slot is what triggers moving the associated RenderObject to a new position. However, in situations involving keyed children as outlined in the bug linked above it is possible that we need to move a RenderObject to a new position in the child list even when its previous sibling hasn't changed.

This change fixes the problem by assigning a new slot value to all children that need their RenderObject moved to trigger the move.

## Related Issues

Fixes https://github.com/flutter/flutter/issues/48855.

## Tests

I added the following tests:

* Render and element tree stay in sync when keyed children move around

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: *Replace with a link to your migration guide*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/widgets/framework.dart [hunk 1]
```diff
@@ -5900,3 +5900,33 @@ FlutterErrorDetails _debugReportException(
...
+class IndexedSlot<T> {
+  /// Creates an [IndexedSlot] with the provided [index] and slot [value].
+  const IndexedSlot(this.index, this.value);
+
+  /// Information to define where the child occupying this slot fits in its
+  /// parent's child list.
+  final T value;
+
+  /// The index of this slot in the parent's child list.
+  final int index;
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/framework.dart]: Now that I look at this again I wonder whether we should just take a generic salt in here instead of an (explicit) index. In either case it probably also makes sense to remove the public getter for that value since nobody should really depend on what the actual value is.

File: packages/flutter/lib/src/widgets/framework.dart [hunk 2]
```diff
@@ -5900,3 +5900,33 @@ FlutterErrorDetails _debugReportException(
...
+
+  @override
+  bool operator ==(Object other) {
+    return other is IndexedSlot
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/framework.dart]: this should compare runtimeType; see style guide for precise template

**Comment by Hixie** [comment]
I don't understand how this solves the problem.

There are a ton of places in our API docs that talk about the slot being an element. We should update all of these and explain in the API docs why that's a bad idea, and how to properly solve the problem, and why this solution works.

**Comment by goderbauer** [comment]
> There are a ton of places in our API docs that talk about the slot being an element.

I found two instances in the API docs (by grepping the docs for "slot") were we make this claim:

https://github.com/flutter/flutter/blob/cc34cb80a829fd22b6d2228e53fe8954d1633b0f/packages/flutter/lib/src/widgets/framework.dart#L5669-L5674

https://github.com/flutter/flutter/blob/cc34cb80a829fd22b6d2228e53fe8954d1633b0f/packages/flutter/lib/src/widgets/framework.dart#L5680-L5683

Agreed, those will need updating.

**Review by chunhtai — DISMISSED** [review]
please ignore

**Comment by goderbauer** [comment]
> I don't understand how this solves the problem.

@Hixie Did my explanation of the problem in the bug (https://github.com/flutter/flutter/issues/48855#issuecomment-592261646) help at all with understanding how/why this fixes the issue? If not, I'l attempt to rephrase the explanation.

**Comment by Hixie** [comment]
> > There are a ton of places in our API docs that talk about the slot being an element.
> 
> I found two instances in the API docs (by grepping the docs for "slot") were we make this claim

Tons, two, you know, same thing... :-)

**Comment by goderbauer** [comment]
Clarified and updated the docs. PTAL.

File: packages/flutter/lib/src/widgets/framework.dart [hunk 3]
```diff
@@ -5783,6 +5809,13 @@ class SingleChildRenderObjectElement extends RenderObjectElement {
 /// RenderObjects use the [ContainerRenderObjectMixin] mixin with a parent data
 /// type that implements [ContainerParentDataMixin<RenderObject>]. Such widgets
 /// are expected to inherit from [MultiChildRenderObjectWidget].
+///
+/// See also:
+///
+/// * [IndexedSlot], which are used as [Element.slot]s for the children of a
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/framework.dart]: s/are/is/

Comment (Hixie) [reply packages/flutter/lib/src/widgets/framework.dart]: or some such. [IndexedSlot] is singular.

File: packages/flutter/lib/src/widgets/framework.dart [hunk 4]
```diff
@@ -5783,6 +5809,13 @@ class SingleChildRenderObjectElement extends RenderObjectElement {
...
+///   [MultiChildRenderObjectElement].
+/// * [RenderObjectElement.updateChildren], which discusses why [IndexedSlot]
+///   are used for the slots of the children.
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/framework.dart]: same grammar issue here

**Comment by Hixie** [comment]
This documentation is fantastic.
LGTM

File: packages/flutter/lib/src/widgets/framework.dart (hunk 2 above)
Comment (chunhtai) [reply packages/flutter/lib/src/widgets/framework.dart]: is ```other is IndexedSlot``` still necessary?

File: packages/flutter/lib/src/widgets/framework.dart [hunk 5]
```diff
@@ -5900,3 +5933,40 @@ FlutterErrorDetails _debugReportException(
   FlutterError.reportError(details);
   return details;
 }
+
+/// Used by [MultiChildRenderObjectElement] for the [Element.slot] value.
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/framework.dart]: The first sentence should be describing what it is. This only describe the usage, maybe rephrase a bit?

File: packages/flutter/lib/src/widgets/framework.dart (hunk 2 above)
Comment (goderbauer) [reply packages/flutter/lib/src/widgets/framework.dart]: Yes, it's necessary for dart to elevate the type of `other` to `IndexedSlot` so I can call `.value` and `.index` below.

**Review by chunhtai — APPROVED** [review]
LGTM
